lib/button.js

```diff
@@ -0,0 +1,326 @@
+    }
+
+    intf.wrapper = wrapper;
+    intf.items = items;
+    init();
+
+    intf.dispose = function () {
+      if (wrapper.parentNode) {
+        wrapper.parentNode.removeChild(wrapper);
+      }
+    };
+    return intf;
+  }
+
+  function toogleButtonMenu() {
+    function createDisposeEvent() {
+      requestAnimationFrame(function () {
+        document.addEventListener('click', self.disposeButtonMenu);
+      });
+    }
+    if (self.buttonMenu) {
+      self.disposeButtonMenu();
+    } else {
+      let pos = {
+        left: self.button.wrapper.left - self.scrollBox.scrollLeft,
+        top:
+          self.button.wrapper.top +
+          self.button.wrapper.offsetHeight -
+          self.scrollBox.scrollTop,
+      };
+      self.buttonMenu = createButtonMenu(pos, self.button.items);
+      self.createInlineStyle(
+        self.button.wrapper,
+        'canvas-datagrid-button-wrapper:active',
+      );
+      createDisposeEvent();
+    }
+  }
+
+  function createButtonMenu(pos, items) {
+    var buttonMenu = document.createElement('div'),
+      selectedIndex = -1,
+      intf = {},
+      rect;
+
+    function createItems() {
+      items.forEach(function (item) {
+        var buttonMenuItemContainer = document.createElement('div');
+
+        function addItem(item) {
+          function addContent(content) {
+            if (content === null) {
+              return;
+            }
+
+            if (typeof content === 'object') {
+              buttonMenuItemContainer.appendChild(content);
+              return;
+            }
+
+            applyButtonMenuItemStyle(buttonMenuItemContainer);
+            buttonMenuItemContainer.innerHTML = content;
+            return;
+          }
+
+          addContent(item.title);
+          item.buttonMenuItemContainer = buttonMenuItemContainer;
+
+          if (item.click) {
+            buttonMenuItemContainer.addEventListener('click', function (ev) {
+              item.click.apply(self, [ev]);
+            });
+          }
+        }
+
+        addItem(item);
+        buttonMenu.appendChild(buttonMenuItemContainer);
+      });
+    }
+
+    function clickIndex(idx) {
+      items[idx].buttonMenuItemContainer.dispatchEvent(new Event('click'));
+    }
+
+    function init() {
+      var loc = {},
+        s = self.scrollOffset(self.canvas);
+
+      if (zIndexTop === undefined) {
+        zIndexTop = self.style.buttonZIndex;
+      }
+
+      createItems();
+      self.createInlineStyle(
+        buttonMenu,
+        'canvas-datagrid-button-menu' + (self.mobile ? '-mobile' : ''),
+      );
+
+      loc.x = pos.left - s.left;
+      loc.y = pos.top - s.top;
+      loc.height = 0;
+      zIndexTop += 1;
+      buttonMenu.style.position = 'absolute';
+      buttonMenu.style.zIndex = zIndexTop;
+      buttonMenu.style.left = loc.x + 'px';
+      buttonMenu.style.top = loc.y + 'px';
+      document.body.appendChild(buttonMenu);
+      rect = buttonMenu.getBoundingClientRect(); // TODO: fix
```

Comment:
Fix what? 😊

Reply:
The comment's gone now, but was there a fix needed, or was there another reason for this comment?

Reply:
sorry, it was my mistake